subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/JavaRecompilationSpecProvider.java

```diff
@@ -119,28 +117,24 @@ private void processOtherChanges(CurrentCompilation current, PreviousCompilation
         boolean emptyAnnotationProcessorPath = current.getAnnotationProcessorPath().isEmpty();
         SourceFileChangeProcessor javaChangeProcessor = new SourceFileChangeProcessor(previous);
         for (FileChange fileChange : sourceFileChanges.get()) {
-            File file = fileChange.getFile();
+            if (spec.isFullRebuildNeeded()) {
+                return;
+            }
```

Comment:
Why do this inside the loop?

Reply:
Since we want to stop early if a full recompilation is needed. Note that `javaChangeProcessor` can set the recompilation reason.